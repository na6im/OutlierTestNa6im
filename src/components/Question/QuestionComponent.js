import React, { useState, useEffect } from 'react'
import {
  Category,
  Question,
  QuestionCore,
  AnswerButton,
  Result
} from './styles'
import fullStar from '../../blackStar.png'
import star from '../../greyStar.png'

function QuestionComponent ({
  question,
  nextQuestionHandler,
  onAnswerHandler,
  end
}) {
  const [answers, setAnswers] = useState([])
  const [answered, setAnswered] = useState()

  const difficulty = {
    easy: 1,
    medium: 2,
    hard: 3
  }

  useEffect(() => {
    let answerOptions = question.incorrect_answers
    answerOptions.splice(
      Math.floor(Math.random() * Math.floor(3)),
      0,
      question.correct_answer
    )
    setAnswers(answerOptions)
    setAnswered()
  }, [question])

  function getAnswerColor (answer) {
    if (answer === question.correct_answer) return 'black'
    if (answer === answered) return 'white'
  }

  function renderResult () {
    return answered === question.correct_answer ? 'Correct' : 'Sorry!'
  }

  function onClickHandler (answer) {
    if (!answered) {
      setAnswered(answer)
      onAnswerHandler(answer === question.correct_answer)
    }
  }

  return (
    <React.Fragment>
      <Category>{decodeURIComponent(question.category)}</Category>
      <div>
        {[...Array(3)].map((x, id) => (
          <img
            alt={'star'}
            key={id}
            src={id < difficulty[question.difficulty] ? fullStar : star}
          />
        ))}
      </div>
      <Question>
        <QuestionCore style={{ gridArea: 'question' }}>
          {decodeURIComponent(question.question)}
        </QuestionCore>
        {answers.map((answer, id) => (
          <AnswerButton
            key={id}
            id={id}
            disabled={
              !!answered &&
              answer !== question.correct_answer &&
              answer !== answered
            }
            onClick={() => {
              onClickHandler(answer)
            }}
            answerColor={!!answered && getAnswerColor(answer)}
          >
            {decodeURIComponent(answer)}
          </AnswerButton>
        ))}
      </Question>
      {!!answered && (
        <Result>
          {!end ? renderResult() : 'its the end'}
          <AnswerButton
            style={{
              width: 'fit-content',
              padding: '5px 1em',
              marginTop: '2em'
            }}
            onClick={() => nextQuestionHandler()}
          >
            {!end ? 'Next question' : 'Play again'}
          </AnswerButton>
        </Result>
      )}
    </React.Fragment>
  )
}

export default QuestionComponent
