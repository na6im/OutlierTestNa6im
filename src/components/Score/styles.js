import styled from 'styled-components'

export const Score = styled.div`
  display: flex;
  flex-direction: column;
  height: 3em;
  width: 100%;
  border-radius: 5px;
  margin-top: auto;
`

export const ScoreStat = styled.div`
  display: flex;
  justify-content: space-between;
  margin-bottom: 5px;
`
export const ScoreBar = styled.div`
  position: relative;
  height: 100%;
  border-radius: 5px;
  overflow: hidden;
  border: solid 1px black;
`
export const Bar = styled.div`
  height: 100%;
  position: absolute;
  left: 0;
  width: ${props => props.ratio * 100}%;
  background-color: black;
`
