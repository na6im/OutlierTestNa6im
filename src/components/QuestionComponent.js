import React, { useState, useEffect } from 'react'
import { Title, Category, Question, QuestionCore, AnswerButton } from './styles'
import star from '../blackStar.png'

function QuestionComponent ({ question }) {
  const [answers, setAnswers] = useState([])
  const [answered, setAnswered] = useState()

  useEffect(() => {
    const answerOptions = question.incorrect_answers
    answerOptions.splice(
      Math.floor(Math.random() * Math.floor(3)),
      0,
      question.correct_answer
    )
    setAnswers(answerOptions)
  }, [])
  {
    //TODO stars
  }

  function getAnswerColor (answer) {
    if (answer === question.correct_answer) return 'black'
    if (answer === answered) return 'white'
  }

  return (
    <React.Fragment>
      <Title>Question 16 of 20</Title>
      <Category>Entertaiment: Board Games</Category>
      <div>
        <img
          src={star}
          style={{ height: '10px', width: '10px', stroke: 'blue' }}
        />
        <img
          src={star}
          style={{ height: '10px', width: '10px', stroke: 'blue' }}
        />
        <img
          src={star}
          style={{ height: '10px', width: '10px', stroke: 'blue' }}
        />
      </div>
      <Question>
        <QuestionCore style={{ gridArea: 'question' }}>
          {decodeURIComponent(question.question)}
        </QuestionCore>
        {answers.map((answer, id) => (
          <AnswerButton
            id={id}
            disabled={
              !!answered &&
              answer !== question.correct_answer &&
              answer !== answered
            }
            onClick={() => setAnswered(answer)}
            answerColor={!!answered && getAnswerColor(answer)}
          >
            {decodeURIComponent(answer)}
          </AnswerButton>
        ))}
      </Question>
    </React.Fragment>
  )
}

export default QuestionComponent
