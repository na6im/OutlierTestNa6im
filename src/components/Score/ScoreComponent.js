import React from 'react'
import { Bar, Score, ScoreBar, ScoreStat } from './styles'

function ScoreComponent ({ score, totalAnswerd, total }) {
  console.log()
  return (
    <Score>
      <ScoreStat>
        <div>
          Score{' '}
          {totalAnswerd === 0 ? 0 : Math.floor((score / totalAnswerd) * 100)}%
        </div>
        <div>
          Max score{' '}
          {Math.floor(((score + (total - totalAnswerd)) / total) * 100)}%
        </div>
      </ScoreStat>
      <ScoreBar>
        <Bar
          ratio={score / total}
          style={{ zIndex: 4, backgroundColor: 'black' }}
        />
        <Bar
          ratio={score / totalAnswerd}
          style={{ zIndex: 3, backgroundColor: '#717171' }}
        />
        <Bar
          ratio={(score + (total - totalAnswerd)) / total}
          style={{ zIndex: 2, backgroundColor: '#D2D2D2' }}
        />
      </ScoreBar>
    </Score>
  )
}

export default ScoreComponent
