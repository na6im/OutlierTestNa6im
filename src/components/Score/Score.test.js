import { render } from '@testing-library/react'
import Score from './ScoreComponent'
import React from 'react'
import ScoreComponent from './ScoreComponent'
import QuestionComponent from '../Question/QuestionComponent'

test('renders corectly', () => {
  const { getAllByTestId, getByText } = render(
    <Score score={1} total={20} totalAnswerd={2} />
  )

  expect(getAllByTestId('bar').length).toBe(3)
  expect(getByText('Score 50%').length)
  expect(getByText('Max score 95%').length)
})
test('Should render and match the snapshot', () => {
  const {
    container: { firstChild }
  } = render(<Score score={1} total={20} totalAnswerd={2} />)
  expect(firstChild).toMatchSnapshot()
})
