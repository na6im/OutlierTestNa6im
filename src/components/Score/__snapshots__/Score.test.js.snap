// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render and match the snapshot 1`] = `
<div
  class="sc-AxjAm hKRLKC"
>
  <div
    class="sc-AxirZ iDpydy"
  >
    <div>
      Score
       
      50
      %
    </div>
    <div>
      Max score
       
      95
      %
    </div>
  </div>
  <div
    class="sc-AxiKw rvBmE"
  >
    <div
      class="sc-AxhCb cqIkFp"
      data-testid="bar"
      style="z-index: 4; background-color: black;"
    />
    <div
      class="sc-AxhCb fOyzjN"
      data-testid="bar"
      style="z-index: 3; background-color: rgb(113, 113, 113);"
    />
    <div
      class="sc-AxhCb jdpbIk"
      data-testid="bar"
      style="z-index: 2; background-color: rgb(210, 210, 210);"
    />
  </div>
</div>
`;
